File: Top_K_Frequently_Repeating_Elements.py
# Time Complexity : O(log n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
# Your code here along with comments explaining your approach
import heapq
class Solution:
    def topKFrequent(self, nums, T):
        d = {}
        for i in range(len(nums)):
            if nums[i] not in d:
                d[nums[i]] = 0 
            d[nums[i]] += 1
        
        pq = []
        for k,v in d.items():
            heapq.heappush(pq, (v, k))
            if len(pq) > T:
                heapq.heappop(pq)
        
        result = [0] * T
        for i in range(len(result)-1, -1, -1):
            curr = heapq.heappop(pq)
            result[i] = curr[1]
        
        return result
    
if __name__ == "__main__":
    s= Solution()
    res = s.topKFrequent([1,1,1,2,2,3], 2)
    assert res == [1,2]

File: TopKFrequentlyRepeatingElements.java
//TC: O(NlogK) -> O(N) if N = K
//SC: O(N+K) -> hashmap size + PriorityQueue size
// Did it run successfully on Leetcode?: Yes
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        // Store the frequency of every number in HashMap
        int res[] = new int[k];
        HashMap<Integer, Integer> map = new HashMap();
        for (int num:nums)
        {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        PriorityQueue minHeap = new PriorityQueue((l1, l2) -> map.get(l1) - map.get(l2));
        
        for(int key: map.keySet())         
        {
            minHeap.offer(key);
            if (minHeap.size() > k)
                // minHeap.poll() always removes the min element from heap. Here it is the element with lowest frequency.
                minHeap.poll();
        }
        // minHeap.poll() always removes the min element from heap, start filling the array from right to left. So left most element in array will have element with highest frequency and right most element of the array will have lowest frequency
        for (int i = k-1; i >= 0; i--)
        {
            res[i] = (int)minHeap.poll();
        }
        return res;
    }
}


